### DISCUSSION TITLE
Using LLDAP over the WAN

### DISCUSSION THREAD

**Discussion by lordratner** [lldap/lldap#618] [Q&A] [answered]
Hello!

I'm setting up a server that is going to be offsite, so it will need to reach out across the internet to my home server to validate via LLDAP.

I just wanted to make sure I am doing this correctly... I assume I should be using SSL, but how does that work? I have NGINX running as a reverse proxy in front of everything, so going to https://password.example.com gets me to the LLDAP web UI using the wildcard certificate for example.com, obviously over port 443 (which NGINX reverse proxies to 17170).

But that's for the webui. How should I be doing this for actually validating a user? I can set up any subdomain, but if it's over port 636 then that would bypass the reverse proxy, and thus the wildcard certificate, right?

I think I need to enable LLDAPS, but that's where I get confused with the certificates. But I don't know if it's possible to use the NGINX Secure connection to then forward to LDAP-insecure so I don't need to enable LLDAPS. 

Obviously I don't know what I'm doing here, but hopefully I explained what I'm *trying* to do? 

Thanks
Seth

**Comment by nitnelave** [comment]
Hi!

In that case, you want to set up LDAPS: check the relevant section in the config, but long story short you give LLDAP the path of the certificates so it can encrypt the LDAP traffic.

**Reply by nitnelave** [reply]
Aaaand I didn't fully read the question, sorry.

It turns out that nginx knows about LDAP, so you can reverse proxy in nginx and handle the certificate in there: https://gist.github.com/jc1518/35cb055bf779f1a70a5fc6e72637407a#file-nginx-conf-L58

**Reply by nitnelave** [reply]
Well, it's not so much that it knows about LDAP, but more that it's able to take an ssl stream and decrypt it before proxying.

**Reply by lordratner** [reply]
Yup, that's what I found too, and it works! I wrote up a bit more detail below if you want to add it as a config/help file.

**Comment by Masgalor** [comment]
Another possibility would be to connect your servers with a vpn-tunnel and use the clear-text-protocol inside this tunnel.

This is especially useful if you have multiple services/ports you want to connect to.
It also reduces the attack-surface by a lot (might be opinionated).

**Comment by lordratner** [comment, marked as answer]
Ok, so I figured this out for my situation, which I suspect might apply to others:

What I'm trying to do: Use the LDAP server on my home server to authenticate for services that are hosted off-site. I want to use the TLS certificates that are already in use for my example.com webserver with minimal changes. 

My setup: I have a virtualized host that uses NGINX, LetsEncrypt, and Authelia in docker (linuxserver.io/SWAG) to reverse proxy all services that have addresses on example.com. This host does not have other services on it (such as LLDAP), so sharing the certs is a bit difficult. Instead of using LDAPS, I want to use a secure connection to the webserver using the example.com certs, then reverse proxy to regular LDAP on my home network. 

The solution: NGINX stream directive, which goes in your nginx.conf file:
```
stream {
    server {
        # LDAPS terminates here
        listen 636 ssl;
        proxy_pass 192.168.10.100:389;   #Your LDAP server IP here, using the unencrypted port
        ssl_certificate /config/keys/cert.crt;  #map to your certs
        ssl_certificate_key /config/keys/cert.key;   #map to your certs

    }
}
```
Note, this needs to be after, not in, the http block

Some things you need to do for this to work:

- Make sure your router forwards port 636 to the webserver
- If using Docker for NGINX, make sure you add port 636 to the docker compose file. Or whatever you use to manage docker.
- When you set up LDAP on whatever service you are using off-site, the LDAP host address must be covered by the domain cert. So example.com, or ldap.example.com, or whatever you set it up with. Also use port 636 (or whatever you set in the stream block in nginx.conf, but not 80/443
- Do NOT set up LDAPS in LLDAP. You can for other uses, but in this case the request goes from NGINX --> LLDAP via unsecured port 389 (or 3890, whatever you're using)

This works if you are using SWAG (which I am). Once you set this up, there is no maintenance, since the certs are handled automatically via  LetsEncrypt.
